File: traveler-app/trekpal/lib/core/utils/validators.dart
// TODO: Implement form validators
// Email, password, CNIC, phone number validation

class AppValidators {
  static String? requiredText(
    String? value, {
    String fieldName = 'This field',
  }) {
    if (value == null || value.trim().isEmpty) {
      return '$fieldName is required';
    }

    return null;
  }

  static String? email(String? value) {
    final String? required = requiredText(value, fieldName: 'Email');
    if (required != null) {
      return required;
    }

    final RegExp emailRegex = RegExp(r'^[^@\s]+@[^@\s]+\.[^@\s]+$');
    if (!emailRegex.hasMatch(value!.trim())) {
      return 'Enter a valid email address';
    }

    return null;
  }

  static String? password(String? value) {
    final String? required = requiredText(value, fieldName: 'Password');
    if (required != null) {
      return required;
    }

    if (value!.trim().length < 8) {
      return 'Password must be at least 8 characters';
    }

    return null;
  }

  static String? cnic(String? value) {
    final String? required = requiredText(value, fieldName: 'CNIC');
    if (required != null) {
      return required;
    }

    final String digitsOnly = value!.replaceAll(RegExp(r'\D'), '');
    if (digitsOnly.length != 13) {
      return 'CNIC must be 13 digits';
    }

    return null;
  }

  static String? minLength(
    String? value, {
    required int min,
    String fieldName = 'This field',
  }) {
    final String? required = requiredText(value, fieldName: fieldName);
    if (required != null) {
      return required;
    }

    if (value!.trim().length < min) {
      return '$fieldName must be at least $min characters';
    }

    return null;
  }

  static String? maxLength(
    String? value, {
    required int max,
    String fieldName = 'This field',
  }) {
    if (value == null || value.trim().isEmpty) {
      return null;
    }

    if (value.trim().length > max) {
      return '$fieldName must be $max characters or less';
    }

    return null;
  }

  static String? phone(String? value, {String fieldName = 'Phone number'}) {
    final String? required = requiredText(value, fieldName: fieldName);
    if (required != null) {
      return required;
    }

    final String digitsOnly = value!.replaceAll(RegExp(r'\D'), '');
    if (digitsOnly.length < 5) {
      return '$fieldName must be at least 5 digits';
    }

    if (value.trim().length > 30) {
      return '$fieldName is too long';
    }

    return null;
  }

  static String? positiveInteger(String? value, {String fieldName = 'Value'}) {
    final String? required = requiredText(value, fieldName: fieldName);
    if (required != null) {
      return required;
    }

    final int? parsed = int.tryParse(value!.trim());
    if (parsed == null || parsed <= 0) {
      return '$fieldName must be greater than zero';
    }

    return null;
  }

  static String? integerLimit(
    String? value, {
    required int min,
    int? max,
    String fieldName = 'Value',
  }) {
    final String? required = requiredText(value, fieldName: fieldName);
    if (required != null) {
      return required;
    }

    final int? parsed = int.tryParse(value!.trim());
    if (parsed == null) {
      return '$fieldName must be a whole number';
    }

    if (parsed < min) {
      return '$fieldName must be at least $min';
    }

    if (max != null && parsed > max) {
      return '$fieldName must be $max or less';
    }

    return null;
  }

  static String? budget(String? value) {
    if (value == null || value.trim().isEmpty) {
      return null;
    }

    final num? parsed = num.tryParse(value.trim());
    if (parsed == null || parsed < 0) {
      return 'Enter a valid budget';
    }

    return null;
  }

  static String? dateOfBirth(DateTime? value) {
    if (value == null) {
      return 'Date of birth is required';
    }

    if (value.isAfter(DateTime.now())) {
      return 'Date of birth cannot be in the future';
    }

    return null;
  }
}

File: vehicle-portal/src/shared/utils/validators.ts
export type ValidationErrors = Record<string, string>;

const emailPattern = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

export const isBlank = (value?: string | null): boolean => !value || !value.trim();

export const validateRequired = (value: string, label: string): string | null => {
  if (isBlank(value)) {
    return `${label} is required`;
  }

  return null;
};

export const validateEmail = (value: string): string | null => {
  if (isBlank(value)) {
    return 'Email is required';
  }

  if (!emailPattern.test(value.trim())) {
    return 'Enter a valid email address';
  }

  return null;
};

export const validatePassword = (value: string): string | null => {
  if (!value) {
    return 'Password is required';
  }

  if (value.length < 8) {
    return 'Password must be at least 8 characters';
  }

  return null;
};

export const validatePhone = (value: string): string | null => {
  if (isBlank(value)) {
    return 'Phone number is required';
  }

  const digits = value.replace(/\D/g, '');
  if (digits.length < 10) {
    return 'Enter a valid phone number';
  }

  return null;
};

export const validateCnic = (value: string): string | null => {
  if (isBlank(value)) {
    return 'CNIC is required';
  }

  if (!/^\d{13}$/.test(value)) {
    return 'CNIC must be exactly 13 digits';
  }

  return null;
};

export const validateMinLength = (
  value: string,
  label: string,
  minLength: number,
): string | null => {
  if (isBlank(value)) {
    return `${label} is required`;
  }

  if (value.trim().length < minLength) {
    return `${label} must be at least ${minLength} characters`;
  }

  return null;
};

export const validatePositiveNumber = (
  value: string | number,
  label: string,
  minValue = 1,
): string | null => {
  const parsedValue = typeof value === 'number' ? value : Number(value);

  if (Number.isNaN(parsedValue)) {
    return `${label} is required`;
  }

  if (parsedValue < minValue) {
    return `${label} must be at least ${minValue}`;
  }

  return null;
};

export const validateFilePresent = (file: File | null, label: string): string | null => {
  if (!file) {
    return `${label} is required`;
  }

  return null;
};

